torchaudio/backend/soundfile_backend.py

```diff
@@ -283,7 +283,7 @@ def _get_subtype(
         if encoding:
             raise ValueError("flac does not support encoding.")
         if not bits_per_sample:
-            return "PCM_24"
+            return "PCM_16"
         if bits_per_sample > 24:
```

Comment:
I wonder if we should also prohibit 24 as it could destroy the signal.
@carolineechen What do you think?

Reply:
I'm not sure I completely understand -- why could using 24-bit destroy the signal?

Reply:
Some process inside of `soundfile` (or `libsndfile` underneath) causes warping from `-1` to `1`. 
https://github.com/pytorch/audio/issues/1592#issuecomment-865141782
I am not sure which and where.

The format is supposed to support up to 32 bits.
https://xiph.org/flac/format.html
> FLAC supports from 4 to 32 bits per sample. 

Using `PCM_16` avoid this problem and `soundfile` defaults to it. IIRC, `soundfile` does not support 32 bit so we raise an error, but for `PCM_24` it does not fail, but gives incorrect result. At least I think we should warn.

Reply:
I see, it looks like it's being caused by something in `libsndfile` where converting from wav to flac creates warping/artifacts ([#274](https://github.com/bastibe/python-soundfile/issues/274), [#265](https://github.com/bastibe/python-soundfile/issues/265)). I think adding a warning makes sense, but I'm not sure about prohibiting 24-bit since it seems this may also be an issue with other bit sizes as well

Reply:
@carolineechen  Thanks for the issue links, then let's add warning and move on here. How does the following message sound?

@yangarbiter Can you add warning something like;

```python
if bits_per_sample == 24:
    warnings.warn("Saving audio with 24 bits per sample might warp samples near -1. Using 16 bits per sample might be able to avoid this.")
```

Reply:
I think the description here is not very precise. Because this distortion not only happens in PCM_24, it also happens in PCM_16. See the below example, the assertion in line 271 will not pass.
![Screen Shot 2021-06-23 at 1 02 14 PM](https://user-images.githubusercontent.com/4192725/123161027-c04abf00-d423-11eb-8c7b-c27017c9a62b.png)

The reason that we can pass the current test is that the assertion here
https://github.com/pytorch/audio/blob/master/test/torchaudio_unittest/backend/soundfile/save_test.py#L277

The variables `expected` and `found` are distorted in the same way.

Because of this, the warning I would suggest is
```
if bits_per_sample == None and ext == "flac":
    warnings.warn("We have changed the default bits_per_sample for flac files from 24 to 16. Please set bits_per_sample to 24 to maintain the same behavior.")
```

Reply:
> Because this distortion not only happens in PCM_24, it also happens in PCM_16

Can you show how it does not pass the assertion? If the reason is absolute difference is slightly larger than the given `atol=1e-4`, then it's fine, because you are saving data in 16 bits. But if the reason is another case of -1 -> 1 warping then it's the real concern.

>     warnings.warn("The default bits_per_sample for flac files has been changed from 24 to 16. Please set bits_per_sample to 24 to maintain the same behavior.")

The core of the concern here is the fact that 24bit could cause warping, and that's what we want users to be aware of it. It's so subtle and difficult to notice.

Reply:
Actually, now that I realized that you are not talking about torch audio's soundfile backend but you are talking about soundfile itself. 

The assertion failure you are mentioning is expected to happen, and that's not how save function is tested. And your modification is invalid for our context. Please refer to http://www.mega-nerd.com/libsndfile/FAQ.html#Q010 for why your assertion is failing.

Reply:
I think this post is talking about small error when you write and read the same file due to float conversion (correct me if I am wrong). Sorry that I should have provide more detail about this. I added a few prints to the code. These error does not seem to be errors due to float conversion.

![Screen Shot 2021-06-24 at 8 29 42 AM](https://user-images.githubusercontent.com/4192725/123290681-5505f900-d4c6-11eb-8e55-0e941cd30006.png)

Here are the print results.
![Screen Shot 2021-06-24 at 8 27 46 AM](https://user-images.githubusercontent.com/4192725/123290919-7ebf2000-d4c6-11eb-9590-6fa271a7eac2.png)

You can see that the fourth and fifth row for the `expected` variable goes completely off. Please let me know if you are also able to reproduce this error.